tensor_mechanics compute=false Material refactor combined with PQPlasticModel refactor

WilkAndy opened the pull request:
Please don't merge yet.

@sapitts and @bwspenc , i'd like your comments on the second of the two commits, please.

This is a small initial exploration towards making my PQPlasticModel a compute=false Material and deriving it from StressUpdateBase.

So far, the important changes are as outlined in the 2nd git commit message, viz:

 

> ComputeReturnMappingStress now has a perform_finite_strain_rotations flag, to save on potentially useless and time-consuming rotations

> ComputeReturnMappingStress now declares plastic_strain.  However, it doesn't compute it: the computation is performed by the classes derived from StressUpdateBase (viz HyperbolicViscoplastcityStressUpdate and IsotropicPlasticityStressUpdate).

> IsotropicPlasticityStressUpdate and HyperbolicViscoplasticityStressUpdate no longer declare the plastic_strain Property.  They have a const_cast on getMaterialProperty in order to update this quantity.

> The reason is I will use lots of plastic models in ComputeReturnMappingStress, but there is only one plastic_strain (up to the "base_name" stuff).  Therefore, it is inappropriate for the individual plastic models to declare their own plastic_strain Property.

> I do not like this design, however, but would like feedback before implementing something else.
What I'd like instead is for plastic_strain to be updated using the inelastic_strain_increments inside ComputeReturnMappingStress::updateQpStress.  This requires the IsotropicPlasticityStressUpdate and HyperbolicViscoplasticityStressUpdate to be slightly refactored.
At the same time, I'd like for RadialReturnStressUpdate::updateStress to also return the consistent tangent operator (which, for IsotropicPlasticityStressUpdate and HyperbolicViscoplasticityStressUpdate will be just the elasticity tensor).  This will allow my derived classes to return the consistent tangent operator that can be combined into Jacobian_mult inside ComputeReturnMappingStress::computeQpStress or ComputeReturnMappingStress::updateQpStress.

> Please let me know what you think of these ideas.

Refs #8561

moosebuild commented:
Job `Precheck:linux-gnu` on 5a1208b wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 1c8e57e38da782bcaff847252e96ff0a04d6d29f`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/8914/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/8914/style.patch | git apply -v`

sapitts commented:
Hi @WilkAndy, I have just a few thoughts at the moment.  I do like the idea of the flag to rotate or not rotate the strains (and I'm a bit concerned that we completely skipped over rotating those strains, now that you've pointed it out).  

The consistent tangent operator would also be a nice addition, although maybe in an class similar to ComputeReturnMappingStress class?  That way the ComputeReturnMappingStress could include @dschwen's isotropic elasticity guarantee.

If ComputeReturnMappingStress will own the plastic_strain variable, then perhaps we should use a more general term; I'll let @bwspenc suggest a preferred name however

WilkAndy commented:
Thanks for your comments @sapitts .   Yes, let's include the isotropic gaurantee of @dschwen so we don't have to rotate the elasticity tensor in ComputeReturnMappingStress.

I'm imaging that IsotropicPlasticityStressUpdate and others will return their consistent tangent operator, ComputeReturnMappingStress will combine all the returned tensors into Jacobian_mult, and finally do the  rotation.  IsotropicPlasticityStressUpdate could also return an "isotropic" gaurantee, so that ComputeReturnMappingStress could figure out whether the combination was also isotropic.

I think "plastic_strain" is a good term actually.  It's pretty universal, isn't it?

WilkAndy commented:
There are 4 real failures in Bison due to moving "plastic_strain".  Are you able to fix Bison if we decide to go ahead with this change, @sapitts or @bwspenc ?   I don't think i'll be ready for this to merge for a couple of weeks, but i don't want to spend all that time and then find out Bison's impossible to fix, or you guys think the plastic_strain (and consistent tangent operator) ideas are stupid!!!

sapitts commented:
Since we also use the ComputeRadialReturnMappingStress to calculate creep, I'd lean towards favoring inelastic_strain, but it's a minor quibble. 

Yes, I can fix Bison, although I'm afraid we might be looking at an incompatible API update, which is always fun.  If it is just the tests that are failing (and not the build), we can just temporarily skip those tests since tensor mechanics in Bison is in more of a sandbox state at present.

sapitts commented:
@acasagran here's the work @WilkAndy is starting on to expand and correct our return mapping algorithm in tensor mechanics; it shouldn't impact your work with the timestep limiter though.

Andy, if you're interested in Al's work, I believe his issue is #8905

WilkAndy commented:
OK, sounds promising about Bison.  It did build ok, but 4 tests failed.

Aha, i see about plastic_strain and inelastic_strain.  The Property was originally called "plastic_strain", but I'm happy to do what you guys think is best, and now's probably a good time to change it.  How about you, @bwspenc ?

I hope i don't impact @acasagran 's work.  i don't think i will.  Anyway, probably his stuff will be merged well before i'm even 1/4 finished!

Overall, thanks for you feedback.  i know it's a bit of a chore to critique other people's work.  While i transition to the compute=false stuff i bet i'll have lots of questions and suggested changes, so thanks for  your input.

a

bwspenc commented on modules/tensor_mechanics/src/materials/TwoStressPlasticity.C:
```diff
@@ -4,17 +4,23 @@
 /*             See LICENSE for full restrictions                */
 /****************************************************************/
-#include "PQPlasticModel.h"
+#include "TwoStressPlasticity.h"
```
I think a more commonly used term would be "Two Invariant Plasticity".  I'd recommend changing the name to that.

WilkAndy replied:
TwoInvariantPlasticity vs TwoStressPlasticity.  I prefer the latter, as the yield function does not need to be a function of a stress invariant.  Indeed, there is already the "WeakPlane" plasticity where the yield function is a function of sigma_zz and a combination of sigma_zx and sigma_zy.  We just need the yield function (and flow potential, etc) to be a function of two independent stress parameters.

bwspenc replied:
OK, fair enough. How about "TwoParameterPlasticity"?  That's another term that seems to be used pretty widely.   (Here's a Wikipedia article that refers to a "three-parameter" model: https://en.wikipedia.org/wiki/Willam-Warnke_yield_criterion).

I just haven't ever heard these models referred to as two (or three) stress models, and a quick web search turns up nothing on that.

WilkAndy replied:
Yep, TwoParameterPlasticity is good, thanks

bwspenc replied:
@WilkAndy arguing about names is what I do best ;)

WilkAndy replied:
I think i'd rename ComuteReturnMappingStress to ComputeAdmissibleConfiguration.  We're not just computing stress, but also inelastic strains and other things, with the requirement that the overall configuration must be admissible.  What do you think @bwspenc and @sapitts ?

Then updateStress could also be called computeAdmissible or something similar.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C:
```diff
@@ -77,10 +92,18 @@ ComputeReturnMappingStress::computeQpStress()
 
+  if (_perform_finite_strain_rotations)
+  {
+    _elastic_strain[_qp] = _rotation_increment[_qp] *
```
The elastic strain and stress already get rotated forward in `ComputeFiniteStrain.C`.  I guess we would need to rotate the plastic strain forward here.

 I'm guessing the motivation for making this rotation optional is to allow for this to work in a small strain setting.  Is that right?  I wonder if there is anything else we need to change for that to work.  We have really focused on finite deformation for these models.

WilkAndy replied:
We inherit from ComputeFiniteStrainElasticStress, which does the rotations on stress, but not elastic strain.  But we override ComputeFiniteStrainElasticStress::computeQpStress anyway, so we have to do the rotation here.

Yes, i wanted to make the rotation optional because of small strains.  i'll keep my eyes out for any trickiness to do with small vs finite strains.

bwspenc replied:
After looking at this again, I think you have the right tensor rotations here.  ComputeFiniteStrain rotates `total_strain` and `mechanical_strain` (which is assigned to `elastic_strain` in ComputeFiniteStrainElasticStress, so you don't have to rotate `elastic_stress` in that model, but you do need to do it here.)

WilkAndy replied:
OK, thanks

bwspenc commented on modules/tensor_mechanics/include/materials/ComputeReturnMappingStress.h:
```diff
@@ -47,12 +48,18 @@ class ComputeReturnMappingStress : public ComputeFiniteStrainElasticStress
   ///@}
 
+  /// plastic strain
+  MaterialProperty<RankTwoTensor> & _plastic_strain;
```
Since this base class is used for both creep and plastic models, it should probably have a more generic name like _inelastic_strain.  We were just dealing with a similar issue for the effective plastic strain (or effective creep strain) variable, and we're still working out a solution.  I like the idea of the base class registering the property, but having the derived class tell the base class what the name of it is.

bwspenc reviewed (commented):
I made a few comments, but this all looks pretty reasonable to me.

moosebuild commented:
Job `Precheck:linux-gnu` on 64580c6 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 583badf8b5e51d7acd3e1f12eed2b76dc4b1dbff`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/8914/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/8914/style.patch | git apply -v`

WilkAndy commented:
@bwspenc and @sapitts .  I'm hoping that while i'm asleep you can review the changes that i just pushed up.  Please read the git commit message.  I'm hoping to work on this tomorrow and need a bit of guidance in places as to what you want.

bwspenc commented on modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C:
```diff
@@ -47,13 +54,22 @@ ComputeReturnMappingStress::ComputeReturnMappingStress(const InputParameters & p
     _elastic_strain_old(declarePropertyOld<RankTwoTensor>(_base_name + "elastic_strain")),
-    _strain_increment(getMaterialProperty<RankTwoTensor>(_base_name + "strain_increment"))
+    _strain_increment(getMaterialProperty<RankTwoTensor>(_base_name + "strain_increment")),
+    _plastic_strain(declareProperty<RankTwoTensor>(_base_name + "plastic_strain")),
```
I think what we decided on here was that we would add a parameter that provide either the name 'plastic' or 'creep', which would get used in the registration of both the 'plastic_strain' and 'effective_plastic_strain' properties in this base class.  You can switch over to using that when we get that sorted out for the effective plastic strain, but how you have it is fine for now.

WilkAndy replied:
OK, thanks

bwspenc replied:
Hold on -- I was mixed up.  This is the class that drives multiple plasticity or creep models, so `_plastic_strain` here contains a combination of the inelastic strains from all of the models that are being driven.  It seems to me like you might want every model to create its own `plastic_strain` or `creep_strain` property in RecomputeRadialReturn, but this one should really have a generic name like `inelastic_strain`.

By the way, I think this class is poorly named, and I've gotten confused like this multiple times. The name `ComputeReturnMappingStress` doesn't convey that this drives multiple models, and the models that it drives don't really even necessarily have to use return mapping algorithms.  I don't have a great alternative to propose at the top of my head, so we'll save renaming it for another time.

WilkAndy replied:
OK, i'll rename to inelastic_strain.  Then the IsotropicPlasticityStressUpdate and HyperbolicViscoplasticityStressUpdate will not be able to use "inelastic_strain", which they currently have.  I'll change theirs to "plastic_strain" for now.

bwspenc commented on modules/tensor_mechanics/tests/recompute_radial_return/affine_plasticity.i:
```diff
@@ -320,8 +320,8 @@
-  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
-  petsc_options_value = 'lu superlu_dist'
+  petsc_options_iname = '-pc_type'
+  petsc_options_value = 'lu'
```
You probably know this, but if you use 'lu', it will only run in serial.

WilkAndy replied:
Yea, i think this test is one element only.  The real reason i removed the superlu_dist was that one of my computers doesn't have it, so the test was failing all the time.

moosebuild commented:
Job `Precheck:linux-gnu` on 1e449fd wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 583badf8b5e51d7acd3e1f12eed2b76dc4b1dbff`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/8914/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/8914/style.patch | git apply -v`

WilkAndy commented:
@bwspenc and @sapitts .  i just pushed another commit.  This adds the initQpInternalProperties, which i believe is necessary to get the plastic_strain MaterialProperty of IsotropicPlasticityStressUpdate (etc) correctly updating.  Please advise if you see anything wrong here.  (See my git commit message also.)

WilkAndy commented:
I've changed some plastic_strain to inelastic_strain.  I'd like to call the MaterialProperty in ComputeReturnMappingStress something like "combined_inelastic_strain", since it is the combination of all the incremental inelastic_strains computed iteratively by the plastic models, however, this breaks tests as the TensorMechanics Action assumes that non-elastic strain is called "plastic_strain".

moosebuild commented:
Job `Precheck:linux-gnu` on 8d492ce wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 583badf8b5e51d7acd3e1f12eed2b76dc4b1dbff`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/8914/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/8914/style.patch | git apply -v`

moosebuild commented:
Job `Precheck:linux-gnu` on d4ddc93 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 583badf8b5e51d7acd3e1f12eed2b76dc4b1dbff`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/8914/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/8914/style.patch | git apply -v`

moosebuild commented:
Job `Precheck:linux-gnu` on 61e8938 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 583badf8b5e51d7acd3e1f12eed2b76dc4b1dbff`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/8914/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/8914/style.patch | git apply -v`

moosebuild commented:
Job `Precheck:linux-gnu` on 51f3aef wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 583badf8b5e51d7acd3e1f12eed2b76dc4b1dbff`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/8914/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/8914/style.patch | git apply -v`

WilkAndy commented:
I had to push this up to github prematurely.  Please don't bother with it @sapitts and @bwspenc .

bwspenc commented:
@WilkAndy I like your proposed name changes for "combined_inelastic_strain" and "ComputeAdmissibleConfiguration".  I know there was some confusion with the current name of that function since it's doing more than just computing stress.

bwspenc commented:
@WilkAndy Changing that function will break a bunch of models in other apps though.  Not that we shouldn't do it -- just warning you.

moosebuild commented:
Job `Precheck:linux-gnu` on c5862ea wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 583badf8b5e51d7acd3e1f12eed2b76dc4b1dbff`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/8914/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/8914/style.patch | git apply -v`

moosebuild commented:
Job `Precheck:linux-gnu` on 035f579 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 583badf8b5e51d7acd3e1f12eed2b76dc4b1dbff`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/8914/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/8914/style.patch | git apply -v`

WilkAndy commented:
@sapitts and @bwspenc .  I've virtually finished this piece of work, and hope to complete it tomorrow.  I need to rename some things (eg computeAdmissibleState) and tidy up the git stuff.

It was very easy to change my PQPlasticModel stuff over to this compute=false Material once i had got the interfaces sorted out.  **However, now my tests run about two times slower!**   Are we sure the compute=false really is not computing stuff?  Or is there another reason for the slowness?  I even made a ComputeReturnMappingStress::updateQpStressSingleModel because i thought maybe all the copying and mucking around with tensors was slowing things down, but no, that made little difference.

permcody commented:
@WilkAndy let us know if you need help tracking down the extra slowdown. I've been having @rwcarlsen run all sorts of profiling lately so if you give us clear instructions on how to run the two tests cases we could narrow this down quickly for you.

WilkAndy committed:
(1) Rename ComputeReturnMappingStress to ComputeAdmissibleState

(2) Rename StressUpdateBase to StateUpdateBase

(3) In ComputeAdmissibleState (used to be ComputeReturnMappingStress):
  - renamed updateQpStress to updateQpState
  - return_mapping_models user parameter renamed to inelastic_models
  - perform_finite_strain_rotations flag added for computational efficiency in small-strain scenarios
  - tangent_operator enum added so user can specify that the full nonlinear consistent tangent operator should be computed, if the plastic models provide that functionality.
  - combined_inelastic_strain Material property added, that defaults to be the sum of the models' inelastic strains, but the user can choose to weight the model contributions using the combined_inelastic_strain_weights parameter
  - finite-strain rotations are performed correctly on elastic strain, plastic strain, Jacobian_mult, etc
  - the models are now passed _elasticity_tensor[_qp], _elastic_strain_old[_qp], _stress_old[_qp], etc, instead of retrieving them as a MaterialProperties.  This is because of the base_name problem: each model could have a different base_name than ComputeAdmissibleState, and that would yield crazy inconsistencies
 - the models are now asked for the consistent tangent operator
 - the combined consistent tangent operator is computed

(4) In StateUpdateBase (used to be StressUpdateBase) minor changes:
 - renamed updateStress to updateState, and this function now includes a lot more parameters, as noted above
 - removed base_name user-input string, as noted above
 - removed any getting of old quantities, due to the potential base_name conflict

(5) RadialReturnStressUpdate minor changes: renaming various Stress to State, as noted above, as well as enhancing the updateState interface

(6) HyperbolicViscoplasticityStressUpdate and IsotropicPlasticityStressUpdate
 - plastic_prepend string added so this class can form its own plastic_strain MaterialProperty without name collisions from other plastic classes
 - various minor changes induced by updateState now including the elasticity_tensor and elastic_strain_old as a parameter instead of this class using getMaterialProperty to get the elasticity tensor

(7) IsotropicPowerLawHardeningStressUpdate and PowerLawCreepStressUpdate and TemperatureDependentHardeningStressUpdate
- various minor changes induced by updateState now including the elasticity_tensor as a parameter instead of this class using getMaterialProperty to get the elasticity tensor

Refs #8561

WilkAndy committed:
(1) Rename PQPlasticModel to TwoParameterPlasticity

(2) In TwoParameterPlasticity
 - now derives from StateUpdateBase
 - Removed declareOld stuff
 - Added name_prepender for the internal plastic things
 - rename updateQpStress to updateState (from StateUpdateBase) and miscellaneous minor interface changes due to this change.  There is no real change to the functionality here.
 - various other minor interface changes, due to rotation_increment, elasticity tensor, stress_old, etc, now coming through updateState rather than getMaterialProperty.

(3) All the classes that derive from TwoParameterPlasticity (ComputeCappedDruckerPragerStress, ComputeCappedWeak*Plane*Stress) have some modified interfaces due to rotation_increment, elasticity tensor, stress_old, etc, now coming through updateState rather than getMaterialProperty.

Refs #8561

WilkAndy committed:
test changes
and miscellaneous changes requested by Stephanie
and resolving merge conflicts
and miscellaneous changes requested by Ben

(1) tests now use ComputeAdmissibleState instead of ComputeReturnMappingStress.

(2) The tests of TwoParameterPlasticity (used to be PQPlasticModel) now use ComputeAdmissibleState

(3) tests/combined_creep_plasticity/combined_creep_plasticity.i utilises the new combined_inelastic_strain_weights parameter

(4) tests/recompute_radial_return/affine_plasticity.i does not use superlu_dist any more

Fixes #8561

WilkAndy commented:
@sapitts and @bwspenc , i've finished with this "compute=false refactor".

I've split the PR into 3 commits, and the first is what will interest you most, i think.  It involves a refactor of your ComputeReturnMappingStress and StressUpdateBase and derived classes.  In the git commit message I've tried to explain the rationale behind the changes.  I don't think there are any major changes.

The second commit involves changing my PQPlasticModel to be a compute=false Material, and enabling it to run with the enhanced stuff of the first commit.

The third commit involves only the test files.  There are no new tests - just slightly modified existing ones.

Please could you review commit (1) ?   You'll have to change Bison, at least, i think.

WilkAndy commented:
@permcody and @rwcarlsen , thanks for your offer of tracking down the slowness caused by the change to a compute=false Material.  I figured it out!

@sapitts and @bwspenc , this is interesting: the test tensor_mechanics/tests/capped_weak_plane/beam.i takes about 6 times longer with perform_finite_strain_rotations=true than it does with perform_finite_strain_rotations=false.  This test does quite a few residual calculations, so does a lot of rotations, and the return-map process itself is fairly cheap in this case.  I presume the main cost is rotating _Jacobian_mult.

WilkAndy commented:
I'm wondering if someone could review this?

permcody commented:
Perhaps @sapitts can take a look?

sapitts commented on modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:
```diff
@@ -12,7 +12,7 @@ template <>
+  InputParameters params = validParams<StateUpdateBase>();
   params.addClassDescription("Calculates the effective inelastic strain increment required to "
                              "return the isotropic stress state to a J2 yield surface.  This class "
                              "is intended to be a parent class for classes with specific "
```
Would you add in a check with Daniels' isotropic tensor guarantee to this class?  Or, since it is very much scope creep, feel free to tag me to complete this check here :-)

WilkAndy replied:
**Not done**.  The problem here is this hair-pulling base_name stuff.  Now RadialReturnStressUpdate doesn't know anything about the elasticity tensor except for through updateState.  It would be too expensive to check the elasticity tensor is isotropic every time updateState is made.  We could add another parameter to updateState (bool elasticity_tensor_is_isotropic) if you like.

sapitts commented on modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C:
```diff
@@ -40,7 +40,7 @@ validParams<RadialReturnStressUpdate>()
+  : StateUpdateBase(parameters),
     _max_its(parameters.get<unsigned int>("max_iterations")),
     _output_iteration_info(getParam<bool>("output_iteration_info")),
     _output_iteration_info_on_error(getParam<bool>("output_iteration_info_on_error")),
```
Suppress the `compute_full_tangent_operator` here?

WilkAndy replied:
**Note done**.  compute_full_tangent operator comes from ComputeAdmissibleState and is not known by StateUpdateBase or derived classes.

WilkAndy replied:
Sorry, should have been **Not done**, not **Note done** !!

sapitts commented on modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h:
```diff
@@ -28,26 +28,35 @@ InputParameters validParams<RadialReturnStressUpdate>();
+                           const RankFourTensor & elasticity_tensor,
+                           const RankTwoTensor & elastic_strain_old,
+                           bool compute_full_tangent_operator,
+                           RankFourTensor & tangent_operator) override;
```
I'm glad to have access to all of these properties for some more complex crystal plasticity models with which I'm working, but I'm a bit concerned about the overhead for the isotropic classes--is it only the beam rotation test that slows down?

WilkAndy replied:
I don't think any of the tests are slower now.  That was a mistake i made with setting perform_finite_strain_rotations=true when i didn't need to.  But it's an interesting observations to keep in mind: when computing the full consistent tangent operator and using finite strains, _Jacobian_mult should be rotated.  This rotation is very computationally intensive.

So the extended interface to updateState has no affect on speed at all.  (And many of the additions to the interface were necessary because of the base_name thing and nothing else.)

sapitts reviewed (commented):
These are pretty much all nit-picky suggestions, that could all be ignored--except the `_cto` variable please.  As always, thank you so much for your great and detailed documentation!

WilkAndy commented:
Thanks @sapitts for your review.  I can implement all/most of your suggestions within the next 3 hours, so i guess this is fairly ready for a BISON patch.  i might not be able to do a lot in the RadialReturn class as i haven't really touched that stuff much (except for changing interfaces) since it's your code.

a

sapitts commented:
Okay, sounds like it's time for me to start muddling through patching BISON then.  I'll fix the RadialReturnMapping stuff in another commit later on

WilkAndy commented on modules/tensor_mechanics/src/materials/ComputeAdmissibleState.C:
```diff
@@ -0,0 +1,281 @@
+
+  if (counter == _max_its && l2norm_delta_stress > _absolute_tolerance &&
+      (l2norm_delta_stress / first_l2norm_delta_stress) > _relative_tolerance)
+    mooseError("Max stress iteration hit during ComputeAdmissibleState solve!");
```
done "throw MooseException"

WilkAndy replied:
done mooseError -> throw MooseException

WilkAndy commented on modules/tensor_mechanics/include/materials/ComputeAdmissibleState.h:
```diff
@@ -0,0 +1,112 @@
+/**
+ * ComputeAdmissibleState computes the stress, the consistent tangent
+ * operator (or an approximation to it), and a decomposition of the strain
+ * into elastic and inelastic parts.  Finite strains are used by default.
```
done used -> assumed

WilkAndy commented on modules/tensor_mechanics/include/materials/ComputeAdmissibleState.h:
```diff
@@ -0,0 +1,112 @@
+                                        RankTwoTensor & combined_inelastic_strain_increment);
+
+  ///@{Input parameters associated with the recompute iteration to return the stress state to the yield surface
+  const unsigned int _max_its;
```
done _max_its -> _max_iterations

WilkAndy commented on modules/tensor_mechanics/include/materials/ComputeAdmissibleState.h:
```diff
@@ -0,0 +1,112 @@
+  const std::vector<Real> _inelastic_weights;
+
+  /// the consistent tangent operators computed by each plastic model
+  std::vector<RankFourTensor> _cto;
```
done _cto -> _consistent_tangent_operator

WilkAndy commented on modules/tensor_mechanics/src/materials/ComputeAdmissibleState.C:
```diff
@@ -0,0 +1,281 @@
+{
+  // Nothing to update during the first time step
+  // ComputeQpStress is not called during the zeroth time step, so no need to guard against _t_step
+  // == 0
```
old comment deleted

WilkAndy commented on modules/tensor_mechanics/src/materials/ComputeAdmissibleState.C:
```diff
@@ -0,0 +1,281 @@
+               << "\n"
+               << " stress convergence relative tolerance = " << _relative_tolerance << "\n"
+               << " absolute l2 norm delta stress = " << l2norm_delta_stress << "\n"
+               << " stress converengen absolute tolerance = " << _absolute_tolerance << std::endl;
```
done converengen -> convergence

acasagran commented:
@WilkAndy I think my recent changes for issue #8905 are causing the conflicts with your merge request. If you need help resolving them, let me know. I think you should be able to just merge my changes in.

WilkAndy commented:
OK, i'll have a look, it shouldn't be too hard, thanks @acasagran

sapitts commented on modules/tensor_mechanics/src/materials/ComputeAdmissibleState.C:
```diff
@@ -0,0 +1,281 @@
+  RankTwoTensor combined_inelastic_strain_increment;
+
+  if (_num_models == 1)
+    updateQpStateSingleModel(elastic_strain_increment, combined_inelastic_strain_increment);
```
@WilkAndy I've got a couple of Bison materials which inherit from RadialReturnStressUpdate but override the StateUpdate method.  When I'm using more than one of these overriding-StateUpdate materials, everything is fine and dandy, but when I use just a single model and use the `updateQpStateSingleModel` I hit a runtime error.  With the debugger the error occurs in libmesh:

```
frame #0: 0x00000001001c26df libbison-dbg.0.dylib`libMesh::TypeVector<libMesh::CompareTypes<double, double>::supertype> libMesh::TypeVector<double>::operator-<double>(this=0x000000010ad9e280, p=0x000000010ad9e7a0) const at type_vector.h:643
   640 	#endif
   641 	
   642 	#if LIBMESH_DIM == 3
-> 643 	  return TypeVector<TS>(_coords[0] - p._coords[0],
   644 	                        _coords[1] - p._coords[1],
   645 	                        _coords[2] - p._coords[2]);
   646 	#endif
```
Haven't hit this error with any of the other 'normal' RadialReturnStressUpdate materials, so it must be something with my odd Bison materials.  I'll keep working on this tomorrow, but if you ran into a similar problem with the other radial return models and could point me in the right direction, I'd greatly appreciate the help.  Thanks!

WilkAndy replied:
No, i haven't had that error.  i don't think it's the "SingleModel" thing, do you?  i think that's just a super stripped-down version of updateQpState, right?

a

WilkAndy replied:
It looks like you're trying to do vector1 - vector2, and either vector1 or vector2 isn't sized to LIBMESH_DIM.  Perhaps this is your axisymmetric thing?

WilkAndy replied:
But probably not with the stress tensor, but probably with a RealVectorValue.   Or maybe a _grad type of variable?

WilkAndy replied:
OK, could it be this (my mistake).  In the not-single case, updateState's second slot (_inelastic_strain_increment[i_rmm]) is zero, but in the single case, it might not be zero(?)   I thought the default constructor would zero it, but maybe not?   All the other classes just set this parameter to something, but perhaps your Bison class uses the parameter before setting it?

sapitts commented:
I can get the classes to run if I comment out the if statement, but, since
other classes run through the single model case, it must be something about
my Bison classes....not sure why the vectors complain.....

On Wed, May 3, 2017, 17:26 Andy Wilkins <notifications@github.com> wrote:

> *@WilkAndy* commented on this pull request.
> ------------------------------
>
> In modules/tensor_mechanics/src/materials/ComputeAdmissibleState.C
> <https://github.com/idaholab/moose/pull/8914#discussion_r114676034>:
>
> > +      mooseError("Model " + models[i] + " is not compatible with ComputeAdmissibleState");
> +  }
> +}
> +
> +void
> +ComputeAdmissibleState::
>
> computeQpStress()
> +{
> +  // Nothing to update during the first time step
> +  // ComputeQpStress is not called during the zeroth time step, so no need to guard against _t_step
> +  // == 0
> +
> +  RankTwoTensor elastic_strain_increment;
> +  RankTwoTensor combined_inelastic_strain_increment;
> +
> +  if (_num_models == 1)
> +    updateQpStateSingleModel(elastic_strain_increment, combined_inelastic_strain_increment);
>
> No, i haven't had that error. i don't think it's the "SingleModel" thing,
> do you? i think that's just a super stripped-down version of updateQpState,
> right?
>
> a
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8914#discussion_r114676034>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI6P5ano5QLWavGk9ijsBJHcrk7_HC8Qks5r2Q0WgaJpZM4M6-3y>
> .
>

sapitts commented:
That might be the case; I'll check tomorrow to see if I can set that
parameter within my Bison class. Thanks!

On Wed, May 3, 2017, 20:17 Andy Wilkins <notifications@github.com> wrote:

> *@WilkAndy* commented on this pull request.
> ------------------------------
>
> In modules/tensor_mechanics/src/materials/ComputeAdmissibleState.C
> <https://github.com/idaholab/moose/pull/8914#discussion_r114690642>:
>
> > +      mooseError("Model " + models[i] + " is not compatible with ComputeAdmissibleState");
> +  }
> +}
> +
> +void
> +ComputeAdmissibleState::computeQpStress()
> +{
> +  // Nothing to update during the first time step
> +  // ComputeQpStress is not called during the zeroth time step, so no need to guard against _t_step
> +  // == 0
> +
> +  RankTwoTensor elastic_strain_increment;
> +  RankTwoTensor combined_inelastic_strain_increment;
> +
> +  if (_num_models == 1)
> +    updateQpStateSingleModel(elastic_strain_increment, combined_inelastic_strain_increment);
>
> OK, could it be this (my mistake). In the not-single case, updateState's
> second slot (_inelastic_strain_increment[i_rmm]) is zero, but in the single
> case, it might not be zero(?) I thought the default constructor would zero
> it, but maybe not? All the other classes just set this parameter to
> something, but perhaps your Bison class uses the parameter before setting
> it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8914#discussion_r114690642>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI6P5Vftr1apIVPoKWx6_quAwOxdE4OEks5r2TUigaJpZM4M6-3y>
> .
>

sapitts commented on modules/tensor_mechanics/src/materials/ComputeAdmissibleState.C:
```diff
@@ -0,0 +1,281 @@
+                          _elasticity_tensor[_qp],
+                          _elastic_strain_old[_qp],
+                          _tangent_operator_type == TangentOperatorEnum::nonlinear,
+                          _Jacobian_mult[_qp]);
```
It turns out that my problematic BISON models desperately need the following four lines right here to work smoothly:
```
  if (_tangent_operator_type == TangentOperatorEnum::elastic)
    _Jacobian_mult[_qp] = _elasticity_tensor[_qp];
  else
    _Jacobian_mult[_qp] = _consistent_tangent_operator[0];
```

Would you pretty please add in these four lines of code please?  Then I will have a working Bison patch ready to go

WilkAndy replied:
I can, but it's suboptimal because it requires copying the RankFourTensor in some cases.  Can you set these in your RadialReturnStressUpdate?  You have the interface

```
updateState(... , elasticity_tensor, ... compute_full_tangent, tangent_operator)
```
so can you just put
```
tangent_operator = elasticity_tensor
```
at the end of your RadialReturnStressUpdate (or whatever class derives from StateUpdateBase)?  That's how it's supposed to work (see around line 170 of RadialReturnStressUpdate)

a

sapitts replied:
Thanks Andy, once I caught all of the early returns in the material, using the tangent operator = elasticity tensor that you suggested worked like a charm without any changes to ComputeAdmissibleState

WilkAndy replied:
Great, i'm glad it all works for you now, @sapitts

sapitts commented:
@bwspenc and @dschwen, would one or both of you take a look through Andy's code and help us get the merge request through please?  I've got a patch for Bison ready and waiting for the moose submodule update that works with Andy's changes.  Thanks!

bwspenc commented:
Sorry -- I've been out a couple of days and was blissfully ignoring this.  I'm reviewing it now.

bwspenc commented on modules/tensor_mechanics/include/materials/ComputeAdmissibleState.h:
```diff
@@ -0,0 +1,112 @@
+/****************************************************************/
```
I hate to make you rename everything, but the more I think about this, the more I don't like this name.  I know that to be precise, the model does much more than compute the stress, but our convention is that all of the models that compute stress have 'Stress' in the name.  We have an awfully large number of models named `Compute*Stress` that compute more than just the stress. 

Also, this name doesn't convey the fact that this is what you use to drive potentially multiple stress update models.  I know that this isn't perfect, but I propose naming this `ComputeMultipleInelasticStress`.  This conveys that it is a driver for potentially multiple inelastic models, and that it is a stress calculator.

bwspenc commented on modules/tensor_mechanics/include/materials/StateUpdateBase.h:
```diff
@@ -0,0 +1,82 @@
+/****************************************************************/
```
Again, since all of the models that derive from this are called `*StressUpdate`, I think we should keep the existing name.

bwspenc commented on modules/tensor_mechanics/include/materials/ComputeCappedWeakPlaneStress.h:
```diff
@@ -7,7 +7,7 @@
 #ifndef COMPUTECAPPEDWEAKPLANESTRESS_H
```
Sorry, but I think we should rename this to `CappedWeakPlaneStressUpdate` to be consistent with the other models.

WilkAndy replied:
done ComputeCappedWeakPlaneStress -> CappedWeakPlaneStressUpdate.  Also similar for ComputeCappedWeakInclinedPlaneStress.  Also similar for TwoParameterPlasticity -> TwoParameterPlasticityStressUpdate

bwspenc commented on modules/tensor_mechanics/include/materials/ComputeCappedDruckerPragerStress.h:
```diff
@@ -7,7 +7,7 @@
 #ifndef COMPUTECAPPEDDRUCKERPRAGERSTRESS_H
```
Now that this is an 'update' model, its name should be named to match our convention: `CappedDruckerPragerStressUpdate.h`

WilkAndy replied:
done ComputeCappedDruckerPragerStress -> CappedDruckerPragerStressUpdate

bwspenc reviewed (changes requested):
The code all looks good.  These names just need to be changed to match the existing conventions.  We should rename ComputeAdmissableState, and then all models that have been converted to be `StressUpdate` models need to be named `*StressUpdate`.

WilkAndy commented:
Thanks @sapitts and @bwspenc for all your comments on the "ComputeAdmissibleState" PR.  I'll probably act on your suggestions on my Monday, so hopefully it'll be ready for you to merge on your Monday.

a

WilkAndy commented on modules/tensor_mechanics/include/materials/TwoParameterPlasticity.h:
```diff
@@ -468,8 +467,9 @@ class PQPlasticModel : public ComputeStressBase
+   * @param rotation_increment The large-strain rotation increment
    */
-  virtual void finaliseReturnProcess();
+  virtual void finaliseReturnProcess(const RankTwoTensor & rotation_increment);
```
done finalise -> finalize

moosebuild commented:
Job `Precheck:linux-gnu` on fd1e2ba wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 4fd5ddeb12dcd26d952b49262a875b54f8cc9e81`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/8914/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/8914/style.patch | git apply -v`

WilkAndy commented:
One of FALCON's tests now fails because it is based on porous_flow which is based on tensor_mechanics.  The fix is simple - just renaming various things.  i'll issue a PR to FALCON when this current PR gets through.

sapitts commented:
Bison patch 1.2 is ready to go now

WilkAndy commented:
Thanks for all your help, everyone.  I'm really pleased with this result and am already using it in real models.

sapitts commented:
@WilkAndy would you like to write an email to the moose users list about the change to the return radial return models?  I've got a lame input file change explanation drafted up, but I'm quite sure you'd explain the mathematical theory behind the input file change much better than I would.  Plus it only seems right that you get the credit for all of the heavy lifting work you did on this commit.--I'll send you and Ben my lame draft and we can go from there?

WilkAndy commented:
​I can do that, but actually this was just a straightforward refactor with no new maths.

a




From: Stephanie Pitts <notifications@github.com>
Sent: Tuesday, 9 May 2017 8:32 AM
To: idaholab/moose
Cc: Wilkins, Andy (Energy, Pullenvale); Mention
Subject: Re: [idaholab/moose] tensor_mechanics compute=false Material refactor combined with PQPlasticModel refactor (#8914)
  
@WilkAndy would you like to write an email to the moose users list about the change to the return radial return models? I've got a lame input file change explanation drafted up, but I'm quite  sure you'd explain the mathematical theory behind the input file change much better than I would. Plus it only seems right that you get the credit for all of the heavy lifting work you did on this commit.--I'll send you and Ben my lame draft and we can go  from there?
—
You are receiving this because you were mentioned.
Reply to this email directly,  view it on GitHub, or  mute the thread.

permcody commented:
All, What's it going to take for Falcon to be patched? This was merged two days ago and no action has been taken. Let's get this fixed up today.

Thanks!

bwspenc commented:
@permcody I just submitted a PR to falcon.  It looks like someone needs to activate the test because I'm not a developer.

WilkAndy commented:
Great, thanks @bwspenc  .  I just woke up and thought i'd better do it now before going to work and getting swamped, and you've already done it!  Cool!
